src/extractor/guidance/sliproad_handler.cpp

```diff
@@ -29,273 +32,730 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
-                    return false;
-                return true;
-            };
+    // The intersection `c` has to be reachable
+    if (!main_road.entry_allowed)
+    {
+        return intersection;
+    }
 
-            if (couldBeSliproad(intersection_following_index_one) &&
-                intersection_node_id != intersection_node_two)
-                return 2;
-            else if (couldBeSliproad(intersection_following_index_two) &&
-                     intersection_node_id != intersection_node_one)
-                return 1;
-            else
-                return 0;
+    // The road leading to the intersection (bc) has to continue from our source
+    if (!roadContinues(source_edge_id, main_road.eid))
+    {
+        return intersection;
+    }
+
+    // Link-check for (bc) and later on (cd) which both are getting shortcutted by Sliproad
+    const auto is_potential_link = [this, main_road](const ConnectedRoad &road) {
+        if (!road.entry_allowed)
+        {
+            return false;
         }
-        else
-            return 0;
-    }();
 
-    if (obvious_turn_index == 0)
-        return intersection;
+        // Prevent from starting in or going onto a roundabout
+        auto onto_roundabout = hasRoundaboutType(road.instruction);
 
-    const auto &next_road = intersection[obvious_turn_index];
+        if (onto_roundabout)
+        {
+            return false;
+        }
 
-    const auto linkTest = [this, next_road](const ConnectedRoad &road) {
-        return !node_based_graph.GetEdgeData(road.eid).roundabout && road.entry_allowed &&
-               angularDeviation(road.angle, STRAIGHT_ANGLE) <= 2 * NARROW_TURN_ANGLE &&
-               !hasRoundaboutType(road.instruction) &&
-               angularDeviation(next_road.angle, road.angle) >
-                   std::numeric_limits<double>::epsilon();
-    };
+        // Narrow turn angle for road (bd) and guard against data issues (overlapping roads)
+        auto is_narrow = angularDeviation(road.angle, STRAIGHT_ANGLE) <= 2 * NARROW_TURN_ANGLE;
+        auto same_angle = angularDeviation(main_road.angle, road.angle) //
+                          <= std::numeric_limits<double>::epsilon();
 
-    bool hasNarrow =
-        std::find_if(intersection.begin(), intersection.end(), linkTest) != intersection.end();
-    if (!hasNarrow)
-        return intersection;
+        if (!is_narrow || same_angle)
+        {
+            return false;
+        }
+
+        const auto road_data = node_based_graph.GetEdgeData(road.eid);
 
-    const auto source_edge_data = node_based_graph.GetEdgeData(source_edge_id);
-    // check whether the continue road is valid
-    const auto check_valid = [this, source_edge_data](const ConnectedRoad &road) {
-        const auto road_edge_data = node_based_graph.GetEdgeData(road.eid);
-        // Test to see if the source edge and the one we're looking at are the same road
+        auto is_roundabout = road_data.roundabout;
```

Comment:
doing roundabout checks early is usually a good idea. Make sure they are up as far as possible in the processing order, since they are relatively cheap compared to walking the graph.